forge/test/mlir/llama/test_llama_inference.py

```diff
@@ -120,3 +121,239 @@ def test_llama_inference_cache_cpu():
+
+    past_key_values_list = [[k, v] for k, v in prefill_output.past_key_values]
+
+    model_inputs = [next_token.unsqueeze(0), past_key_values_list]
+
+    # Padding on past key values can be enabled by setting apply_padding_on_past_key_values to True.
+    apply_padding_on_past_key_values = True
+    max_new_tokens = 32
+    if apply_padding_on_past_key_values:
+        padding_seq_len = max_new_tokens
+        non_padding_seq_len = past_key_values_list[0][0].shape[-2]
+
+        # Zero Pad past key values in key_value_seq_len(i.e -2) dimension
+        # Before padding past key values tensor shape -> (batch_size, num_of_key_values_heads, key_value_seq_len, head_dim)
+        # After Padding Past key value tensor shape -> (batch_size, num_of_key_values_heads, key_value_seq_len + padding_seq_len, head_dim)
+        for idx, (k, v) in enumerate(model_inputs[1]):
+            model_inputs[1][idx][0] = torch.cat(
+                [
+                    k,
+                    torch.zeros(
+                        k.shape[-4], k.shape[-3], max_new_tokens, k.shape[-1]
+                    ).to(k.dtype),
+                ],
+                dim=-2,
+            )
+            model_inputs[1][idx][1] = torch.cat(
+                [
+                    v,
+                    torch.zeros(
+                        v.shape[-4], v.shape[-3], max_new_tokens, v.shape[-1]
+                    ).to(k.dtype),
+                ],
+                dim=-2,
+            )
+
+    # # Compile the model
+    # compiled_model = forge.compile(llama_model, sample_inputs=model_inputs)
+
+    # Run decode stage on TT device and generate tokens by passing the last predicted token and the past key values.
+    # untill the a specified maximum number of new tokens is reached or an end-of-sequence token is encountered.
+    for max_new_tokens_idx in range(max_new_tokens):
+
+        # CPU Inference
+        model_outputs = llama_model(model_inputs[0], model_inputs[1])
+
+        # TT will return the logits and past key values as list of tensor, so flattening
+        # framework output past key values from List(List(Key1, Values1), ... , List(Key26, Values26)) to
+        # List(Key1, Values1, ... , Key26, Values26) for comparing the Framework and TT output in similar fashion.
+        framework_output = [model_outputs[0]]
+        for k, v in model_outputs[1]:
+            framework_output.append(k)
+            framework_output.append(v)
+
+        # # Run on TT device
+        # tt_output = compiled_model(*model_inputs)
+        # tt_output = [tt_out.to("cpu") for tt_out in tt_output]
+
+        # # Validate TT result with Framework
+        # assert all([compare_with_golden_pcc(golden=fw_out, calculated=tt_out, pcc=0.99) for fw_out, tt_out in zip(framework_output, tt_output)])
+
+        logits = framework_output[0]
+        past_key_values = framework_output[1:]
+
+        next_token_logits = logits[:, -1, :]
```

Comment:
Next token should be calculated based on logits on index `non_padding_seq_len + max_new_tokens_idx`, not last -1. Because last logits correspond to padding tokens in this case. This would be ok if we didn't have padding. Then, last token would be on -1 but now it isn't.